### PR TITLE
feat: add new route matching param position

### LINKED ISSUE #1961: support add more parameters in advanced routing matching conditions
# Feature request

## Please describe your feature

We want to add matching rules for ngx.var.server_port , but it is currently not supported in advanced routing matching conditions

## Describe the solution you'd like

Add a new parameter position，so that we can customize the parameters

### PR DESCRIPTION
**Pull request by lookbrook**
Please answer these questions before submitting a pull request, **or your PR will get closed**.

**Why submit this pull request?**

- [ ] Bugfix
- [x] New feature provided
- [ ] Improve performance
- [ ] Backport patches

**What changes will this PR take into?**

add a new route matching param position, so that we can use the build-in parameter such as var.server_port

**Related issues**

fix/resolve #1961

**Checklist:**

- [x] Did you explain what problem does this PR solve? Or what new features have been added?
- [ ] Have you added corresponding test cases?
- [ ] Have you modified the corresponding document?
- [ ] Is this PR backward compatible? If it is not backward compatible, please discuss on the mailing list first

### CODE REVIEW

File: web/src/pages/Route/components/Step1/MatchingRulesView.tsx [hunk 1]
```diff
@@ -104,6 +104,9 @@ const MatchingRulesView: React.FC<RouteModule.Step1PassProps> = ({
           case 'cookie':
             renderText = 'Cookie';
             break;
+          case 'buildin':
```
Comment (juzhiyuan) [inline web/src/pages/Route/components/Step1/MatchingRulesView.tsx:107]: Hi, usually we will call it `builtin`, could you please provide a link on this param?

**Comment by juzhiyuan** [comment]
image

see here, need your help to update codes :)

File: web/src/pages/Route/components/Step1/MatchingRulesView.tsx (hunk 1 above)
Comment (lookbrook) [reply web/src/pages/Route/components/Step1/MatchingRulesView.tsx:107]: This is the documentation for the relevant embedded parameters: http://nginx.org/en/docs/http/ngx_http_core_module.html#variables

File: web/src/pages/Route/components/Step1/MatchingRulesView.tsx [hunk 2]
```diff
@@ -217,6 +220,7 @@ const MatchingRulesView: React.FC<RouteModule.Step1PassProps> = ({
               <Option value="http">{formatMessage({ id: 'page.route.httpRequestHeader' })}</Option>
               <Option value="arg">{formatMessage({ id: 'page.route.requestParameter' })}</Option>
               <Option value="cookie">Cookie</Option>
+              <Option value="buildin">{formatMessage({ id: 'page.route.buildinParameter' })}</Option>
```
Comment (juzhiyuan) [inline web/src/pages/Route/components/Step1/MatchingRulesView.tsx:223]: need @liuxiran and @tzssangglass 's confirmation.

[1] https://apisix.apache.org/docs/apisix/admin-api#request-body-parameters
[2] https://github.com/api7/lua-resty-expr

Comment (tzssangglass) [reply web/src/pages/Route/components/Step1/MatchingRulesView.tsx:223]: cookie is builtin Nginx variables.

"cookie_x_foo" means to use the key with "x_foo" cookie.

Comment (juzhiyuan) [reply web/src/pages/Route/components/Step1/MatchingRulesView.tsx:223]: Just to make sure, if there has an Option called `buildin` here?

Comment (tzssangglass) [reply web/src/pages/Route/components/Step1/MatchingRulesView.tsx:223]: I think `buildin` is required, see: http://nginx.org/en/docs/varindex.html

Comment (juzhiyuan) [reply web/src/pages/Route/components/Step1/MatchingRulesView.tsx:223]: ok

**Comment by liuxiran** [comment]
Hi @lookbrook , the code looks good to me, would you please add the related test case for this nice feature?

you can refer to https://github.com/apache/apisix-dashboard/blob/9b18c35d3c61ccfb8c48bd458568c86ac902491e/web/cypress/integration/route/create-edit-duplicate-delete-route.spec.js#L105

to create a route with `buildin vars` and configure it, check if this rule seted successfully, and then change it to another param, check if it can configure successfully.

**Comment by lookbrook** [comment]
> Hi @lookbrook , the code looks good to me, would you please add the related test case for this nice feature?
> 
> you can refer to
> 
> https://github.com/apache/apisix-dashboard/blob/9b18c35d3c61ccfb8c48bd458568c86ac902491e/web/cypress/integration/route/create-edit-duplicate-delete-route.spec.js#L105
> 
> to create a route with `buildin vars` and configure it, check if this rule seted successfully, and then change it to another param, check if it can configure successfully.

ok, I will try it

**Review by liuxiran — APPROVED** [review]

**Review by juzhiyuan — APPROVED** [review]

**Comment by liuxiran** [comment]
also cc @Baoyuantop @iamayushdas @guoqqqi for review, thanks a lot

**Review by iamayushdas — APPROVED** [review]

**Review by guoqqqi — APPROVED** [review]

File: web/cypress/integration/route/create-route-with-advanced-matching-conditions.spec.js [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+ */
+/* eslint-disable no-undef */
+
+context('Create Route with advanced matching conditions', () => {
+  const selector = {
+    name: '#name',
+    nodes_0_host: '#nodes_0_host',
+    nodes_0_port: '#nodes_0_port',
+    nodes_0_weight: '#nodes_0_weight',
+    input: ':input',
```
Comment (guoqqqi) [inline web/cypress/integration/route/create-route-with-advanced-matching-conditions.spec.js:25]:
```suggestion
    input: ':input',
```
The dom element does not appear to use

**Comment by guoqqqi** [comment]
LGTM!

**Review by bzp2010 — APPROVED** [review]

**Pull request merged**